File: AtCoder/ABC/193/D.cpp
#pragma GCC optimize("O3")
#include <iostream>
#include <iomanip>
#include <cstdio>
#include <string>
#include <cstring>
#include <deque>
#include <list>
#include <queue>
#include <stack>
#include <vector>
#include <utility>
#include <algorithm>
#include <map>
#include <set>
#include <complex>
#include <cmath>
#include <limits>
#include <cfloat>
#include <climits>
#include <ctime>
#include <cassert>
#include <numeric>
#include <fstream>
#include <functional>
#include <bitset>
using namespace std;

using ll = long long;
using P = pair<int, int>;
using T = tuple<int, int, int>;

template <class T> inline T chmax(T &a, const T b) {return a = (a < b) ? b : a;}
template <class T> inline T chmin(T &a, const T b) {return a = (a > b) ? b : a;}

constexpr int MOD = 1e9 + 7;
constexpr int inf = 1e9;
constexpr long long INF = 1e18;

#define all(a) (a).begin(), (a).end()

int dx[] = {1, 0, -1, 0};
int dy[] = {0, 1, 0, -1};

int main(){
    cin.tie(0);
    ios::sync_with_stdio(false);

    int k; cin>>k;
    string s, t; cin>>s>>t;

    vector<int> rest(10, k);
    vector<int> tak(10, 0), aok(10, 0);
    for(int i=0; i<4; i++){
        rest[s[i] - '0']--, rest[t[i] - '0']--;
        tak[s[i] - '0']++, aok[t[i] - '0']++;
    }

    double ans = 0;
    for(int i=1; i<=9; i++){
        for(int j=1; j<=9; j++){
            if(i == j && rest[i] < 2) continue;
            if(rest[i] < 1 || rest[j] < 1) continue;

            int tsum = 0;
            for(int l=1; l<=9; l++){
                int cur = l;
                for(int m=0; m<tak[l]; m++) cur *= 10;
                if(l == i) cur *= 10;
                tsum += cur;
            }

            int asum = 0;
            for(int l=1; l<=9; l++){
                int cur = l;
                for(int m=0; m<aok[l]; m++) cur *= 10;
                if(l == j) cur *= 10;
                asum += cur;
            }

            if(tsum > asum){
                ans += 1. * rest[i] * (rest[j] - (i == j)) / (9 * k - 8) / (9 * k - 9);
            }
        }
    }

    printf("%.10f\n", ans);

    return 0;
}

File: AtCoder/ABC/193/D.py
k = int(input())
s = input()
t = input()

rest = [k for i in range(10)]
tak = [0 for i in range(10)]
aok = [0 for i in range(10)]

for i in range(4):
    rest[int(s[i])] -= 1
    rest[int(t[i])] -= 1
    tak[int(s[i])] += 1
    aok[int(t[i])] += 1

ans = 0
for x in range(1, 10):
    for y in range(1, 10):
        if rest[x] < 1 or rest[y] < 1 or (x == y and rest[x] < 2):
            continue
    
        tak[x] += 1
        aok[y] += 1

        tsum = 0
        asum = 0
        for i in range(1, 10):
            tsum += i * 10 ** tak[i]
            asum += i * 10 ** aok[i]

        tak[x] -= 1
        aok[y] -= 1

        if tsum > asum:
            ans += rest[x] * (rest[y] - (x == y)) / (9 * k - 8) / (9 * k - 9)

print(ans)